admin-manual/security/csp-headers.rst

```diff
@@ -0,0 +1,263 @@
+
+   Adding a script's hash to the CSP directive allows for specific inline scripts
+   to execute based on their hashed content. While this method ensures only the
+   approved script runs, it's fragile because any change to the script's content
+   requires recalculating and updating its hash in the CSP. This can be cumbersome
+   and error-prone, especially with frequent script modifications.
+
+   In addition, if there are a large number of violations, there could mean a lot of
+   hashes to add, creating a very messy CSP directive.
+
+3. For a given request generate a random nonce value to be returned in the CSP
+   header directive and on each inline script, style etc tag.
+
+   Each inline script or style has to be flagged as 'ok' with a nonce (a random
+   token) value that is uniquely generated for a specific request/response pair.
+   This nonce must be included in the Header directive, and in each inline script
+   (or style, font, img source) tag. If your application has lots of inline
+   scripts, styles and whatnot this could be a lot of work, but it will be
+   more robust than using script hashes.
+
+   .. code-block:: none
+
+      Content-Security_policy-Report-Only: default-src 'self'; font-src 'self'; img-src 'self' https://www.gravatar.com/avatar/ blob:; script-src 'self' 'nonce-abcd1234567890'; style-src 'self' 'nonce-abcd1234567890'; worker-src 'self' blob:; frame-ancestors 'self';
+
+4. Remove the inline asset! Refactor the code to move the inline asset to the
+   script and style bundle. If it's an inline style can it be replaced with a
+   Bootstrap 5 equivalent?
+
+Work completed to make the Dominion theme compatible with CSP headers can be viewed
+in this `AtoM CSP commit`_. This commit provides examples of how to refer to the nonce
+value generated by AtoM from your theme templates, and examples of refactoring code to 
+remove inline styles in favour of Bootstrap 5 equivalents.
+
+Allowing External Sources
+-------------------------
+
+If for example the application makes use of Gravatar assets, we could allow them by 
+adding:
+
+.. code-block:: none
+
+   Content-Security-Policy-Report-Only: default-src 'self'; font-src 'self'; img-src 'self' https://www.gravatar.com/avatar/; script-src 'self'; style-src 'self'; frame-ancestors 'self';
+
+Note it is good to be as specific as possible without creating a maintainance
+headache when specifying the domain. E.g. if all assets are loaded from the
+``avatar`` endpoint, then it is better to be more specific than allow scripts
+to be run from the entire ``https://www.gravatar.com`` domain.
+
+If your theme makes use of Google Analytics, Tag Manager, or the Maps API, then
+you may need to whitelist additional sources. We recommend consulting Google's
+documentation for this:  
+
+* https://developers.google.com/tag-manager/web/csp
+* https://developers.google.com/web/fundamentals/security/csp/
+* https://content-security-policy.com/examples/google-maps/
+
+.. SEEALSO::
```

Comment:
Nice!

Reply:
I think this was from your security doc! :+1: